rewrite numpy array hook and matrix statistics with c-api

lijun99 opened the pull request:
1) rewrote numpy array reference (asnumpy) for gsl matrix/vector with numpy c api, 
defined in extensions/gsl/numpy.cc 
2) rewrote statistics (mean/sd) with numpy.mean style 
array.mean(axis) where axis = None (flattened), 0 (along row), 1 (along column). 
for standard deviation, an option sample=True(False) is added to differentiate 
sample standard deviation ( normalization factor N-1) and population sd (normalization factor N).

aivazis commented on extensions/gsl/numpy.cc:
```diff
@@ -0,0 +1,95 @@
+const char * const gsl::vector::asnumpy__doc__ = "return a numpy array reference of vector";
+
+PyObject *
+gsl::vector::asnumpy(PyObject *, PyObject * args) {
```
should we name this "ndarray" instead of "asnumpy"?

aivazis commented on extensions/gsl/numpy.cc:
```diff
@@ -0,0 +1,95 @@
+const char * const gsl::matrix::asnumpy__doc__ = "return a numpy array reference of matrix";
+
+PyObject *
+gsl::matrix::asnumpy(PyObject *, PyObject * args) {
```
should we name this "ndarray" instead of "asnumpy"?

EJFielding commented:
My build of pyre with the old mm failed:

> std-cpp.def:47: warning: undefined variable `NUMPY_INCDIR'
g++ -MMD -DCXX_ID="gcc-4.x-Darwin" -DDEBUG -DBLD_PROCEDURE -DWITH_GSL -DHAVE_INLINE -DWITH_NUMPY -DWITH_MPI -DPARALLEL -DWITH_PARALLEL_IO -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/config/platform/Darwin-10.x -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/config/compiler/gcc-4.x-Darwin -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/config/compiler/gcc-4.x-Darwin -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/config/target/Darwin-10.x_x86_64 -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/products/include -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/products/include -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/products/include -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/products/include -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/products/include -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre//products/include -I/opt/local/include -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m -I/opt/local/include/openmpi-gcc7  -Wall  -g        -std=c++17 -fPIC   -c -o /Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/builds/Darwin-10.x_x86_64-debug/tmp/extensions/gsl/numpy.o numpy.cc
numpy.cc:18:10: fatal error: numpy/arrayobject.h: No such file or directory
 #include <numpy/arrayobject.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.

lijun99 commented:
Hi Eric,

Indeed, the new numpy support needs additional change to
pyre/.mm/config.def
which is machine dependent and not synced to repo.

You will need to add the following line to config.def
NUMPY_INCDIR =
opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/numpy/core/include
(for python3.6 on macports, which I guess you are using).
For other versions and/or platforms, you may find the numpy include path by
import numpy; numpy.get_include()
and add the result as NUMPY_INCDIR to config.def.

-Lijun


On Thu, Mar 14, 2019 at 9:59 PM Eric Fielding <notifications@github.com>
wrote:

> My build of pyre with the old mm failed:
>
> std-cpp.def:47: warning: undefined variable `NUMPY_INCDIR'
> g++ -MMD -DCXX_ID="gcc-4.x-Darwin" -DDEBUG -DBLD_PROCEDURE -DWITH_GSL
> -DHAVE_INLINE -DWITH_NUMPY -DWITH_MPI -DPARALLEL -DWITH_PARALLEL_IO
> -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/config/platform/Darwin-10.x
> -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/config/compiler/gcc-4.x-Darwin
> -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/config/compiler/gcc-4.x-Darwin
> -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/config/target/Darwin-10.x_x86_64
> -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/products/include
> -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/products/include
> -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/products/include
> -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/products/include
> -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/products/include
> -I/Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre//products/include
> -I/opt/local/include
> -I/opt/local/Library/Frameworks/Python.framework/Versions/3.6/include/python3.6m
> -I/opt/local/include/openmpi-gcc7 -Wall -g -std=c++17 -fPIC -c -o
> /Users/fielding/Documents/progs/Caltech/extra_aivazis/pyre/builds/Darwin-10.x_x86_64-debug/tmp/extensions/gsl/numpy.o
> numpy.cc
> numpy.cc:18:10: fatal error: numpy/arrayobject.h: No such file or directory
> #include <numpy/arrayobject.h>
> ^~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pyre/pyre/pull/18#issuecomment-473159342>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Ak8ZGe-BJyYOVaJ_6PjEQ-aTU4CkVA2Sks5vWyitgaJpZM4bkDNp>
> .
>

EJFielding commented:
Thanks. I have my .mm/config.def in my home directory on my Mac, but I changed it there and now I can build pyre. I guess the installation notes need to be updated for this additional requirement.

lijun99 commented:
Just did that. I also updated python 3.7 as the recommended version,
which new pyre updates require.

-Lijun

On Fri, Mar 15, 2019 at 8:41 AM Eric Fielding <notifications@github.com> wrote:
>
> Thanks. I have my .mm/config.def in my home directory on my Mac, but I changed it there and now I can build pyre. I guess the installation notes need to be updated for this additional requirement.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

EJFielding commented:
Do we need to install Python 3.7 to run anything with pyre? I saw that one of the tests failed with something that apparently is in Python 3.7 but not 3.6. It is quite a lot of work to reinstall all of my MacPorts Python modules for the new version.

lijun99 commented:
python 3.6 should work well for now (except one pyre test). But I recommend
any new user to start with python 3.7.

- Lijun

On Fri, Mar 15, 2019 at 9:17 AM Eric Fielding <notifications@github.com>
wrote:

> Do we need to install Python 3.7 to run anything with pyre? I saw that one
> of the tests failed with something that apparently is in Python 3.7 but not
> 3.6. It is quite a lot of work to reinstall all of my MacPorts Python
> modules for the new version.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pyre/pyre/pull/18#issuecomment-473349416>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Ak8ZGbsypkEzECXYyG65e6Y-C8BZqBzvks5vW8eLgaJpZM4bkDNp>
> .
>
